File: server/controllers/ResumeController.js
const { default: mongoose } = require('mongoose');
const Resume = require('../model/Resume')
const cloudinary = require('cloudinary').v2
const auth = require('../middlewares/auth');

// Configure Cloudinary once at module load
cloudinary.config({
    cloud_name: process.env.CLOUDINARY_CLOUD_NAME,
    api_key: process.env.CLOUDINARY_API_KEY,
    api_secret: process.env.CLOUDINARY_API_SECRET,
})

class ResumeController {
    static async createNewResume(req, res) {
        try {
            const { _id, userId, ...data } = req.body; // ignore incoming userId
            const auth = req.auth || {};
            if (auth.type !== 'user' || !auth.userId) return res.status(401).json({ message: 'Unauthorized' });

            const newResume = new Resume({ ...data, userId: auth.userId });
            await newResume.save();

            return res.status(201).json({ message: 'Created successfully!', newResume });
        } catch (err) {
            return res.status(500).json({ message: err.message });
        }
    }

    static async deleteResume(req, res) {
        try {
            const authCtx = req.auth || {};
            if (authCtx.type !== 'user' || !authCtx.userId) {
                return res.status(401).json({ message: 'Unauthorized' });
            }

            const { id } = req.params;
            if (!id || !mongoose.Types.ObjectId.isValid(id)) {
                return res.status(400).json({ message: 'Invalid resume id' });
            }

            const deleted = await Resume.findOneAndDelete({ _id: id, userId: authCtx.userId });
            if (!deleted) {
                return res.status(404).json({ message: 'Resume not found' });
            }

            return res.status(200).json({ message: 'Deleted successfully', id });
        } catch (err) {
            console.error('Error deleting resume:', err);
            return res.status(500).json({ message: err.message });
        }
    }

    static async duplicate(req, res) {
        try {
            const { _id, userId, ...data } = req.body; //Định nghĩa _id để loại _id cũ
            const auth = req.auth || {};
            if (auth.type !== 'user' || !auth.userId) return res.status(401).json({ message: 'Unauthorized' });

            const newResume = new Resume({ ...data, userId: auth.userId });
            await newResume.save();

            return res.status(200).json({ 
                message: "Created a duplicate.",
                newResume: newResume
            })

        } catch (err) {
            return res.status(500).json({ message: err.message })
        }
    }

    static async saveResume(req, res) {
        try {
            const { _id, userId, ...data } = req.body;
            const auth = req.auth || {};
            if (auth.type !== 'user' || !auth.userId) return res.status(401).json({ message: 'Unauthorized' });

            // Update existing resume by _id
            if (_id) {
                if (!mongoose.Types.ObjectId.isValid(_id)) {
                    return res.status(400).json({ message: 'Invalid resume id' });
                }

                const updated = await Resume.findOneAndUpdate(
                    { _id, userId: auth.userId },
                    { ...data, updatedAt: Date.now() },
                    { new: true }
                );

                if (!updated) {
                    return res.status(404).json({ message: 'Resume not found' });
                }

                return res.status(200).json({ message: 'Updated successfully', resume: updated });
            }

            // Create new resume if no _id provided
            const userObjectId = new mongoose.Types.ObjectId(auth.userId);

            const newResume = new Resume({
                ...data,
                userId: userObjectId,
            });
            await newResume.save();

            return res.status(201).json({ message: 'Saved successfully', resume: newResume });
        } catch (err) {
            console.error('Error saving resume:', err);
            return res.status(500).json({ message: err.message });
        }
    }

    // static async getResume(req, res) {
    //     try {
    //         // const { userId } = req.params
    //         const userId = '68f10f356594cb494a29067e'; // ID bạn muốn thử
    //         const userObjectId = new mongoose.Types.ObjectId(userId);
    //         // if (mongoose.Types.ObjectId.isValid(userId)) {
                
    //         //     console.log(userObjectId);
    //         // } else {
    //         //     console.log('Invalid ObjectId');
    //         // }

    //         const resume = await Resume.findOne({ userId: userObjectId })

    //         if (!resume) {
    //             return res.status(404).json({ message: 'Not found' });
    //         }

    //         return res.status(200).json({ resume });

    //     } catch (err) {
    //         console.error('Error during resume fetching:', err);
    //         return res.status(500).json({ message: err.message });
    //     }
    // }
    
    // static async getResume(req, res) {
    //     try {
    //         // const { id } = req.params; // nếu lấy từ URL /api/resume/:id
    //         const resumeId = "68f4f57056a71e0214ca0b05"; // _id bạn muốn thử
    //         const objectId = new mongoose.Types.ObjectId(resumeId);

    //         // Tìm theo _id thay vì userId
    //         const resume = await Resume.findById(objectId);

    //         if (!resume) {
    //         return res.status(404).json({ message: "Resume not found" });
    //         }

    //         return res.status(200).json({ resume });
    //     } catch (err) {
    //         console.error("Error during resume fetching:", err);
    //         return res.status(500).json({ message: err.message });
    //     }
    // }

    static async getResumesByUserId(req, res) {
        try {
            const auth = req.auth || {};
            if (auth.type !== 'user' || !auth.userId) return res.status(401).json({ message: 'Unauthorized' });
            const userObjectId = new mongoose.Types.ObjectId(auth.userId);

            const resumes = await Resume.find({ userId: userObjectId })
                                        .sort({ createdAt: -1 });

            if (!resumes || resumes.length === 0) {
                return res.status(200).json({
                    resumes: [],
                    message: "No resumes yet"
                });

            }

            return res.status(200).json({ resumes });

        } catch (err) {
            console.error('Error fetching resumes by userId:', err);
            return res.status(500).json({ message: err.message });
        }


    }

    // Uploads a base64/dataURL image to Cloudinary and returns the URL
    static async uploadAvatar(req, res) {
        try {
            const { image } = req.body || {};
            if (!image || typeof image !== 'string') {
                return res.status(400).json({ message: 'Missing image' });
            }

            const folder = process.env.CLOUDINARY_FOLDER || 'resume-avatar';

            const result = await cloudinary.uploader.upload(image, {
                folder,
                resource_type: 'image',
                overwrite: true,
            });

            return res.status(200).json({ url: result.secure_url });
        } catch (err) {
            console.error('Error uploading avatar:', err);
            return res.status(500).json({ message: err.message });
        }
    }

}

module.exports = ResumeController

File: server_2/controllers/resume_controller.py
import os
from datetime import datetime
from typing import Any, Dict, List, Optional

import cloudinary
import cloudinary.uploader
from bson import ObjectId
from fastapi import HTTPException, status
from motor.motor_asyncio import AsyncIOMotorDatabase
from pymongo import ReturnDocument

cloudinary.config(
    cloud_name=os.getenv("CLOUDINARY_CLOUD_NAME"),
    api_key=os.getenv("CLOUDINARY_API_KEY"),
    api_secret=os.getenv("CLOUDINARY_API_SECRET"),
)

TEMPLATE_MAP = {"classic": "Classic", "modern": "Modern", "minimalist": "Minimalist"}


def normalize_template(value: Optional[str]) -> str:
    if not value or not isinstance(value, str):
        return "Classic"
    key = value.strip().lower()
    return TEMPLATE_MAP.get(key, "Classic")


def serialize_resume(doc: Dict[str, Any]) -> Dict[str, Any]:
    """Convert Mongo types to JSON-friendly variants."""
    serialized = {**doc}
    if "_id" in serialized:
        serialized["_id"] = str(serialized["_id"])
    if "userId" in serialized:
        serialized["userId"] = str(serialized["userId"])
    for key in ("createdAt", "updatedAt", "share", "basics", "styles", "sections"):
        # leave complex structures as-is; Mongo primitives are JSON serializable
        pass
    return serialized


def _ensure_object_id(value: str) -> ObjectId:
    try:
        return ObjectId(value)
    except Exception as exc:  # pylint: disable=broad-except
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Invalid resume id") from exc


async def create_new_resume(db: AsyncIOMotorDatabase, user_id: str, payload: Dict[str, Any]) -> Dict[str, Any]:
    resume_data = {k: v for k, v in payload.items() if k not in {"_id", "userId"}}
    resume_data["userId"] = _ensure_object_id(user_id)
    resume_data["template"] = normalize_template(payload.get("template"))
    now = datetime.utcnow()
    resume_data.setdefault("createdAt", now)
    resume_data.setdefault("updatedAt", now)
    resume_data.setdefault("share", {"enabled": False})
    resume_data.setdefault("visibility", "private")

    result = await db.resumes.insert_one(resume_data)
    created = await db.resumes.find_one({"_id": result.inserted_id})
    return serialize_resume(created)


async def delete_resume(db: AsyncIOMotorDatabase, user_id: str, resume_id: str) -> str:
    oid = _ensure_object_id(resume_id)
    deleted = await db.resumes.find_one_and_delete({"_id": oid, "userId": _ensure_object_id(user_id)})
    if not deleted:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Resume not found")
    return str(oid)


async def duplicate_resume(db: AsyncIOMotorDatabase, user_id: str, payload: Dict[str, Any]) -> Dict[str, Any]:
    clean_fields = {k: v for k, v in payload.items() if k not in {"_id", "userId", "share", "visibility", "createdAt", "updatedAt"}}
    clean_fields["userId"] = _ensure_object_id(user_id)
    clean_fields["template"] = normalize_template(payload.get("template"))
    clean_fields["share"] = {"enabled": False}
    clean_fields["visibility"] = "private"
    clean_fields["createdAt"] = datetime.utcnow()
    clean_fields["updatedAt"] = datetime.utcnow()

    result = await db.resumes.insert_one(clean_fields)
    created = await db.resumes.find_one({"_id": result.inserted_id})
    if not created:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Failed to duplicate resume")
    return serialize_resume(created)


async def save_resume(db: AsyncIOMotorDatabase, user_id: str, payload: Dict[str, Any]) -> Dict[str, Any]:
    resume_id = payload.get("_id")
    normalized_template = None
    if "template" in payload:
        normalized_template = normalize_template(payload.get("template"))

    if resume_id:
        oid = _ensure_object_id(resume_id)
        update_payload = {k: v for k, v in payload.items() if k not in {"_id", "userId"}}
        update_payload["updatedAt"] = datetime.utcnow()
        if normalized_template:
            update_payload["template"] = normalized_template

        updated = await db.resumes.find_one_and_update(
            {"_id": oid, "userId": _ensure_object_id(user_id)},
            {"$set": update_payload},
            return_document=ReturnDocument.AFTER,
        )
        if not updated:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Resume not found")
        return serialize_resume(updated)

    # Create new resume
    resume_data = {k: v for k, v in payload.items() if k not in {"_id", "userId"}}
    resume_data["userId"] = _ensure_object_id(user_id)
    resume_data["template"] = normalized_template or normalize_template(None)
    resume_data["createdAt"] = datetime.utcnow()
    resume_data["updatedAt"] = datetime.utcnow()
    result = await db.resumes.insert_one(resume_data)
    created = await db.resumes.find_one({"_id": result.inserted_id})
    return serialize_resume(created)


async def get_resumes_by_user(db: AsyncIOMotorDatabase, user_id: str) -> List[Dict[str, Any]]:
    cursor = db.resumes.find({"userId": _ensure_object_id(user_id)}).sort("createdAt", -1)
    results: List[Dict[str, Any]] = []
    async for doc in cursor:
        results.append(serialize_resume(doc))
    return results


async def upload_avatar(image: str, folder_env: str = "CLOUDINARY_FOLDER") -> str:
    if not image or not isinstance(image, str):
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Missing image")

    folder = os.getenv(folder_env, "resume-avatar")
    result = cloudinary.uploader.upload(
        image,
        folder=folder,
        resource_type="image",
        overwrite=True,
    )
    return result.get("secure_url") or ""
